### ISSUE TITLE
Flow Interprets Type Spread as Possibly Undefined

### ISSUE THREAD

**Issue by arahansen** [facebook/flow#5054]
Flow seems to incorrectly state that a value could be possibly undefined when using type spread

```js
// @flow

type OtherProps = {
  user: {
    name: string,
  }
}

type Props = {
  food: {
    amount: number,
  },
}

// flow says `user` could possibly be undefined, even though type says user is always defined
function Tester(props: { ...Props, ...OtherProps }) {
  return props.user.name
}

// using the intersection syntax, flow can interpret the types correctly
function Tester2(props: Props & OtherProps) {
  return props.user.name 
}
```

example on [flow.org/try](https://flow.org/try/#0PTAEAEDMBsHsHcBQiAuBPADgU1AeRQBZYBOACsbBgM6gC8oA3oqKAK5UkBcjzLoAdgEMAtlm5UUxAJb8A5gBpeAX0QrUmHOUo16TFpFiwAJtz18RsVvxTd+rYQCMSilksVqQoGAlBVBaGgADdhJA0ABjS2gjUAxYKiopB2g0UCc2fiMsSBksI3lQLAA3LH5QQktZAnKNX38aEOJQKRpBaHh60Cyc-jzESCtwlClYMoAVLAkSAAoMCmpTUAA6Fa1qApWl-CIyeZolAEoeFmIsFFZiMrntJcaloVFVZE92GVlyombrEg4hkbKqGhrIIAB4FbzwCKCMoyFAkOZnD44dDYGiRYinIYpfqDYajUATKbEABMsz23DWNAAZHhCCRKUczKdzpdYntbhxiPcRDgnkA)

I get the same error if I define `type Props` as:
```js
type Props = {
  ...OtherProps,
  food: {
    amount: number,
  },
}

// should expect type checking to pass in the same manner as the previous example
// but this fails
function Tester(props: Props) {
  return props.user.name
}
```

**Comment by AugustinLF** [comment]
If you use exact types, it works! See this [example](https://flow.org/try/#0PTAEAEDMBsHsHcBQiAuBPADgU1AeRQBZYBOACsbBgM6gC8oA3gD6KigCuVJAXI626AB2AQwC2WXlRTEAloIDmAGn4BfREzWpMOcpRr0G-SLFgATXoYGgxsdoJS9B7UQCMSytiuWaQoGAlAqYTQaAANOElDQAGNbaFNQDFgqKhkXaDRQNw5BUyxIOSxTRVAsADcsQVBCW3kCau1A4JoI4lAZGmFoeGbQPILBIsRIO2iUGVgqgBUsKRIACgwKagtQADoN3WoSjbX8IjJlmhUASj42YiwUdmIqpb011rWRcUQfME45eWqidvsSLhjCZVKhoezCAAeJX88BiwiqchQJCWVx+OHQ2BosWIlzGGWGo3Gk1AMzmxAATIsjrwtjQAGR4QgkWlnSygS7XW6JI6PLjEZ5iHBvRBAA). The problem is that flow considers that OtherProps could have a food property :)

**Comment by arahansen** [comment]
@AugustinLF thanks for the example! This still seems to me like a bug within flow though. Or is this intended behavior?

**Comment by asolove** [comment]
I think this is the expected behavior. The type spread operator is meant to have the same left to right semantics as object spread in js. Because the second object in your spread is non exact, it is allowed to have extra properties, including ones that conflict with the first spread object. As a result, flow doesn’t know for sure what type that key will have. By using an exact object for the second object, flow knows it won’t have extra keys and won’t override the first one.

**Comment by ianobermiller** [comment]
@asolove I don't think that logic holds for either of the two examples above. In the first, `OtherProps`, with `user` is being spread **after** `Props`, so Flow should know that `user` cannot be undefined. In the second case, `food` is explicitly defined after the spread, so Flow should also know that it cannot be undefined. 

If we're talking about `food` being potentially undefined this logic does make sense. Even so, using Exact types does help, good call!
